# Shell Lab's Novozymes Competition Files
## Contents
- **databases**
  - contains the training and test datasets provided by Novozymes for the Kaggle competition.
- **regression**
  - trains and tests a regression model for predicting Tm from sequence
  - It has a separate README.md. Check it out!
- **clustering**
  - contains output from [MMseqs clustering](https://academic.oup.com/bioinformatics/article/32/9/1323/1744460) of training dataset protein sequences for which we have Tm measurements, PDB files from AlphaFold + the PDB that corresponds to training set sequence clusters, and code for procuring these structures.
  - Subdirectories:
    - **clusterSeqs** - FASTA files of all sequence clusters generated by MMseqs.
    - **subtraining_clusterPDBs** - PDB structures for centroid sequences of clusters and code for pulling these from the PDB code:
      - *pdb_structure_retriever.py*
      - *pdb_batch_download.sh*
- **tools**
  - Various Python files with functions to extract features from sequences and structures and train the ML model
  - Code:
    - *sequence_analysis.py*
    - *map_structure_to_cluster.py* - contains functions that compute sequence + structural features of training dataset sequences
    - *gen_subtraining_model.py* - uses map_structure_to_cluster.py functions to create dataset for ML model creation + training
    - *regress.py* - trains + tests a regression model for predicting Tm from sequence
- **energies_from_ambertools**
  - Contains .csv files with various energies for each test set pdb (**energies_test_set.csv**) and each training set pdb (**energies_training.csv**) that were output by the protein language model [EMBER3D](https://github.com/kWeissenow/EMBER3D).
  - Energies (columns of .csv files):
    - e_angle, e_bond, e_dihedral, e_gb (generalized born), e_elec, e_elec14, e_vdw, e_vdw14, e_tot
  - Code:
    - *compute_amber_energy.py* & *compute_amber_energies_batches.py* - uses ambertools to calculate energies on Pod. It references Kevin's folder on pod with all the EMBER outputted pdbs (`/home/kshen/Share/novozyme/novozymes-competition/EMBER/`)
    - *submit.sh* and *submit_batches.sh* - Pod submission scripts
    - *min.in* - input file for energy minimization that is referenced by the python scripts. It does 10 steps of steepest descent and 10 steps of conjugate gradient minimization. You can reduce these numbers to increase the speed of the code.

